File: frontend/mobile-app/ios/WaqitiApp/CertificatePinningModule.swift
import Foundation
import Security
import CommonCrypto

/**
 * Native module for certificate pinning on iOS
 * Provides comprehensive SSL/TLS security for network communications
 */
@objc(CertificatePinningModule)
class CertificatePinningModule: NSObject, URLSessionDelegate {
    
    private var pinnedSession: URLSession?
    private var domainPins: [String: [String]] = [:]
    private var enforceMode: Bool = true
    private var certificateCache: [String: SecCertificate] = [:]
    
    // Production certificate pins for Waqiti services
    private let defaultPins: [String: [String]] = [
        "api.example.com": [
            "sha256/BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB=",
            "sha256/CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC=" // Backup pin
        ],
        "auth.example.com": [
            "sha256/DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD=",
            "sha256/EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE=" // Backup pin
        ],
        "wallet.example.com": [
            "sha256/FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF=",
            "sha256/GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG=" // Backup pin
        ]
    ]
    
    override init() {
        super.init()
        self.domainPins = defaultPins
        initializePinnedSession()
    }
    
    /**
     * Initialize URLSession with certificate pinning
     */
    private func initializePinnedSession() {
        let configuration = URLSessionConfiguration.default
        
        // Configure TLS settings
        configuration.tlsMinimumSupportedProtocolVersion = .TLSv12
        configuration.tlsMaximumSupportedProtocolVersion = .TLSv13
        
        // Security settings
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        configuration.urlCache = nil
        configuration.httpShouldUsePipelining = false
        configuration.httpShouldSetCookies = false
        
        // Timeout settings
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 60
        
        // Create session with pinning delegate
        pinnedSession = URLSession(
            configuration: configuration,
            delegate: self,
            delegateQueue: nil
        )
        
        print("[CertificatePinning] Session initialized with \(domainPins.count) pinned domains")
    }
    
    /**
     * Configure pinned session with custom pins
     */
    @objc func configurePinnedSession(
        _ config: NSDictionary,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard let domains = config["domains"] as? [[String: Any]] else {
            rejecter("CONFIG_ERROR", "Invalid configuration format", nil)
            return
        }
        
        domainPins.removeAll()
        
        for domain in domains {
            guard let domainName = domain["domain"] as? String,
                  let pins = domain["pins"] as? [String] else {
                continue
            }
            domainPins[domainName] = pins
        }
        
        if let enforce = config["enforceMode"] as? Bool {
            enforceMode = enforce
        }
        
        // Reinitialize session with new configuration
        initializePinnedSession()
        
        resolver([
            "success": true,
            "domainCount": domainPins.count
        ])
    }
    
    /**
     * Test certificate pinning for a specific domain
     */
    @objc func testPinning(
        _ hostname: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard let url = URL(string: "https://\(hostname)/health") else {
            rejecter("URL_ERROR", "Invalid hostname", nil)
            return
        }
        
        let task = pinnedSession?.dataTask(with: url) { data, response, error in
            if let error = error {
                // Check if it's a pinning failure
                if (error as NSError).code == NSURLErrorServerCertificateUntrusted {
                    resolver([
                        "success": false,
                        "hostname": hostname,
                        "error": "PINNING_FAILED",
                        "message": error.localizedDescription
                    ])
                    self.sendSecurityEvent(type: "pinning_test_failed", details: hostname)
                } else {
                    rejecter("TEST_ERROR", error.localizedDescription, error)
                }
            } else {
                resolver([
                    "success": true,
                    "hostname": hostname,
                    "pinned": self.domainPins[hostname] != nil
                ])
            }
        }
        
        task?.resume()
    }
    
    /**
     * Get current pinning status
     */
    @objc func getStatus(
        _ resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        var domainInfo: [String: Int] = [:]
        for (domain, pins) in domainPins {
            domainInfo[domain] = pins.count
        }
        
        resolver([
            "enabled": pinnedSession != nil,
            "enforceMode": enforceMode,
            "domainCount": domainPins.count,
            "domains": domainInfo
        ])
    }
    
    /**
     * Update certificate pins dynamically
     */
    @objc func updatePins(
        _ updates: NSDictionary,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard let domains = updates["domains"] as? [[String: Any]] else {
            rejecter("UPDATE_ERROR", "Invalid update format", nil)
            return
        }
        
        for domain in domains {
            guard let hostname = domain["hostname"] as? String,
                  let pins = domain["pins"] as? [String] else {
                continue
            }
            domainPins[hostname] = pins
        }
        
        // Reinitialize session with updated pins
        initializePinnedSession()
        
        sendSecurityEvent(
            type: "pins_updated",
            details: "Updated \(domains.count) domains"
        )
        
        resolver([
            "success": true,
            "updatedDomains": domains.count
        ])
    }
    
    /**
     * Reset to default pins
     */
    @objc func resetToDefaults(
        _ resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        domainPins = defaultPins
        initializePinnedSession()
        
        resolver([
            "success": true,
            "defaultDomains": defaultPins.count
        ])
    }
    
    // MARK: - URLSessionDelegate
    
    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        // Only handle server trust challenges
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }
        
        let host = challenge.protectionSpace.host
        
        // Check if we have pins for this host
        guard let pins = domainPins[host] ?? domainPins[getBaseDomain(host)] else {
            if enforceMode {
                // No pins configured, reject if in enforce mode
                completionHandler(.cancelAuthenticationChallenge, nil)
                sendSecurityEvent(type: "no_pins_for_host", details: host)
            } else {
                // Allow connection without pinning
                completionHandler(.performDefaultHandling, nil)
            }
            return
        }
        
        // Validate certificate chain
        if validateCertificateChain(serverTrust: serverTrust, pins: pins, host: host) {
            let credential = URLCredential(trust: serverTrust)
            completionHandler(.useCredential, credential)
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            sendSecurityEvent(type: "pinning_validation_failed", details: host)
        }
    }
    
    /**
     * Validate certificate chain against pins
     */
    private func validateCertificateChain(
        serverTrust: SecTrust,
        pins: [String],
        host: String
    ) -> Bool {
        // Perform standard certificate validation
        var error: CFError?
        let isValid = SecTrustEvaluateWithError(serverTrust, &error)
        
        if !isValid {
            print("[CertificatePinning] Standard validation failed for \(host): \(error?.localizedDescription ?? "Unknown error")")
            return false
        }
        
        // Get certificate chain
        let certificateCount = SecTrustGetCertificateCount(serverTrust)
        
        for i in 0..<certificateCount {
            guard let certificate = SecTrustGetCertificateAtIndex(serverTrust, i) else {
                continue
            }
            
            // Calculate pin for this certificate
            let pin = calculatePin(for: certificate)
            
            // Check if pin matches any of our configured pins
            if pins.contains(pin) {
                print("[CertificatePinning] Pin matched for \(host)")
                return true
            }
        }
        
        print("[CertificatePinning] No matching pins found for \(host)")
        return false
    }
    
    /**
     * Calculate SHA256 pin for a certificate
     */
    private func calculatePin(for certificate: SecCertificate) -> String {
        guard let publicKey = SecCertificateCopyKey(certificate),
              let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, nil) as Data? else {
            return ""
        }
        
        // Calculate SHA256 hash
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        publicKeyData.withUnsafeBytes { bytes in
            _ = CC_SHA256(bytes.baseAddress, CC_LONG(publicKeyData.count), &hash)
        }
        
        // Convert to base64
        let hashData = Data(hash)
        return "sha256/" + hashData.base64EncodedString()
    }
    
    /**
     * Get base domain from hostname
     */
    private func getBaseDomain(_ host: String) -> String {
        let components = host.split(separator: ".")
        if components.count > 2 {
            // Return last two components (e.g., "waqiti.com" from "api.example.com")
            return components.suffix(2).joined(separator: ".")
        }
        return host
    }
    
    /**
     * Send security events to JavaScript
     */
    private func sendSecurityEvent(type: String, details: String) {
        guard let bridge = self.bridge else { return }
        
        bridge.eventDispatcher().sendDeviceEvent(
            withName: "CertificatePinningSecurityEvent",
            body: [
                "type": type,
                "details": details,
                "timestamp": Date().timeIntervalSince1970 * 1000
            ]
        )
    }
    
    // MARK: - React Native Module Setup
    
    @objc static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    @objc var methodQueue: DispatchQueue {
        return DispatchQueue(label: "com.waqiti.certificatepinning", qos: .userInitiated)
    }
}

// MARK: - Objective-C Bridge

@objc(CertificatePinningModule)
extension CertificatePinningModule: RCTBridgeModule {
    static func moduleName() -> String! {
        return "CertificatePinningModule"
    }
}

// MARK: - Export Methods to React Native

extension CertificatePinningModule {
    @objc func exportedMethods() -> [String] {
        return [
            "configurePinnedSession",
            "testPinning",
            "getStatus",
            "updatePins",
            "resetToDefaults"
        ]
    }
}

File: frontend/mobile-app/android/app/src/main/java/com/waqiti/mobile/CertificatePinningModule.java
package com.waqiti.mobile;

import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.io.IOException;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLPeerUnverifiedException;
import javax.net.ssl.SSLSession;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.CertificatePinner;
import okhttp3.ConnectionSpec;
import okhttp3.OkHttpClient;
import okhttp3.TlsVersion;

/**
 * Native module for certificate pinning on Android
 * Provides comprehensive SSL/TLS security for network communications
 */
public class CertificatePinningModule extends ReactContextBaseJavaModule {

    private static final String TAG = "CertificatePinning";
    private static final String MODULE_NAME = "CertificatePinningModule";
    
    private final ReactApplicationContext reactContext;
    private OkHttpClient pinnedClient;
    private CertificatePinner certificatePinner;
    private Map<String, List<String>> domainPins;
    private boolean enforceMode = true;
    
    // Production certificate pins for Waqiti services
    // THREAD-SAFETY & SECURITY FIX: Use immutable collections to prevent runtime modifications
    private static final Map<String, List<String>> DEFAULT_PINS;

    static {
        Map<String, List<String>> tempPins = new HashMap<>();
        tempPins.put("api.example.com", Collections.unmodifiableList(Arrays.asList(
            "sha256/BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB=",
            "sha256/CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC=" // Backup pin
        )));
        tempPins.put("auth.example.com", Collections.unmodifiableList(Arrays.asList(
            "sha256/DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD=",
            "sha256/EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE=" // Backup pin
        )));
        tempPins.put("wallet.example.com", Collections.unmodifiableList(Arrays.asList(
            "sha256/FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF=",
            "sha256/GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG=" // Backup pin
        )));
        // CRITICAL: Make entire map immutable to prevent certificate pin tampering
        DEFAULT_PINS = Collections.unmodifiableMap(tempPins);
    }

    public CertificatePinningModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.domainPins = new HashMap<>(DEFAULT_PINS);
        initializePinnedClient();
    }

    @Override
    public String getName() {
        return MODULE_NAME;
    }

    /**
     * Initialize OkHttpClient with certificate pinning
     */
    private void initializePinnedClient() {
        try {
            // Build certificate pinner
            CertificatePinner.Builder pinnerBuilder = new CertificatePinner.Builder();
            
            for (Map.Entry<String, List<String>> entry : domainPins.entrySet()) {
                String domain = entry.getKey();
                for (String pin : entry.getValue()) {
                    pinnerBuilder.add(domain, pin);
                    pinnerBuilder.add("*." + domain, pin); // Include subdomains
                }
            }
            
            certificatePinner = pinnerBuilder.build();
            
            // Configure TLS versions and cipher suites
            ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
                .tlsVersions(TlsVersion.TLS_1_3, TlsVersion.TLS_1_2)
                .cipherSuites(
                    "TLS_AES_256_GCM_SHA384",
                    "TLS_AES_128_GCM_SHA256",
                    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
                    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
                )
                .build();
            
            // Build OkHttpClient with certificate pinning
            OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder()
                .certificatePinner(certificatePinner)
                .connectionSpecs(Arrays.asList(spec, ConnectionSpec.COMPATIBLE_TLS))
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .hostnameVerifier(new StrictHostnameVerifier());
            
            // Add custom trust manager for additional validation
            if (enforceMode) {
                clientBuilder.sslSocketFactory(
                    createSecureSocketFactory(),
                    new PinningTrustManager()
                );
            }
            
            pinnedClient = clientBuilder.build();
            
            Log.i(TAG, "Certificate pinning initialized successfully");
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize certificate pinning", e);
            sendSecurityEvent("pinning_initialization_failed", e.getMessage());
        }
    }

    /**
     * Configure pinned client with custom pins
     */
    @ReactMethod
    public void configurePinnedClient(ReadableMap config, Promise promise) {
        try {
            // Parse configuration
            ReadableArray hosts = config.getArray("hosts");
            boolean enforce = config.hasKey("enforceMode") ? 
                config.getBoolean("enforceMode") : true;
            
            this.enforceMode = enforce;
            this.domainPins.clear();
            
            // Add pins for each host
            for (int i = 0; i < hosts.size(); i++) {
                ReadableMap host = hosts.getMap(i);
                String hostname = host.getString("hostname");
                ReadableArray pins = host.getArray("pins");
                
                List<String> pinList = new ArrayList<>();
                for (int j = 0; j < pins.size(); j++) {
                    pinList.add(pins.getString(j));
                }
                
                domainPins.put(hostname, pinList);
            }
            
            // Reinitialize client with new pins
            initializePinnedClient();
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putInt("hostCount", domainPins.size());
            promise.resolve(result);
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to configure pinned client", e);
            promise.reject("CONFIG_ERROR", "Failed to configure certificate pinning", e);
        }
    }

    /**
     * Test certificate pinning for a specific domain
     */
    @ReactMethod
    public void testPinning(String hostname, Promise promise) {
        try {
            // Attempt to connect to the hostname
            okhttp3.Request request = new okhttp3.Request.Builder()
                .url("https://" + hostname + "/health")
                .head()
                .build();
            
            okhttp3.Response response = pinnedClient.newCall(request).execute();
            response.close();
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putString("hostname", hostname);
            result.putBoolean("pinned", domainPins.containsKey(hostname));
            
            promise.resolve(result);
            
        } catch (SSLPeerUnverifiedException e) {
            // Certificate pinning failed
            Log.e(TAG, "Certificate pinning failed for " + hostname, e);
            
            WritableMap error = Arguments.createMap();
            error.putBoolean("success", false);
            error.putString("hostname", hostname);
            error.putString("error", "PINNING_FAILED");
            error.putString("message", e.getMessage());
            
            sendSecurityEvent("pinning_test_failed", hostname);
            promise.resolve(error);
            
        } catch (Exception e) {
            Log.e(TAG, "Error testing pinning for " + hostname, e);
            promise.reject("TEST_ERROR", "Failed to test certificate pinning", e);
        }
    }

    /**
     * Get current pinning status
     */
    @ReactMethod
    public void getStatus(Promise promise) {
        try {
            WritableMap status = Arguments.createMap();
            status.putBoolean("enabled", pinnedClient != null);
            status.putBoolean("enforceMode", enforceMode);
            status.putInt("domainCount", domainPins.size());
            
            WritableMap domains = Arguments.createMap();
            for (Map.Entry<String, List<String>> entry : domainPins.entrySet()) {
                domains.putInt(entry.getKey(), entry.getValue().size());
            }
            status.putMap("domains", domains);
            
            promise.resolve(status);
            
        } catch (Exception e) {
            promise.reject("STATUS_ERROR", "Failed to get pinning status", e);
        }
    }

    /**
     * Update certificate pins dynamically
     */
    @ReactMethod
    public void updatePins(ReadableMap updates, Promise promise) {
        try {
            // Verify updates structure
            ReadableArray domains = updates.getArray("domains");
            
            for (int i = 0; i < domains.size(); i++) {
                ReadableMap domain = domains.getMap(i);
                String hostname = domain.getString("hostname");
                ReadableArray pins = domain.getArray("pins");
                
                List<String> pinList = new ArrayList<>();
                for (int j = 0; j < pins.size(); j++) {
                    pinList.add(pins.getString(j));
                }
                
                // Update or add domain pins
                domainPins.put(hostname, pinList);
            }
            
            // Reinitialize client with updated pins
            initializePinnedClient();
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putInt("updatedDomains", domains.size());
            
            sendSecurityEvent("pins_updated", "Updated " + domains.size() + " domains");
            promise.resolve(result);
            
        } catch (Exception e) {
            Log.e(TAG, "Failed to update pins", e);
            promise.reject("UPDATE_ERROR", "Failed to update certificate pins", e);
        }
    }

    /**
     * Clear all custom pins and revert to defaults
     */
    @ReactMethod
    public void resetToDefaults(Promise promise) {
        try {
            domainPins.clear();
            domainPins.putAll(DEFAULT_PINS);
            initializePinnedClient();
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putInt("defaultDomains", DEFAULT_PINS.size());
            
            promise.resolve(result);
            
        } catch (Exception e) {
            promise.reject("RESET_ERROR", "Failed to reset pins", e);
        }
    }

    /**
     * Get the configured OkHttpClient for use by other modules
     */
    public OkHttpClient getPinnedClient() {
        return pinnedClient;
    }

    /**
     * Create secure SSL socket factory
     */
    private javax.net.ssl.SSLSocketFactory createSecureSocketFactory() throws Exception {
        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, new TrustManager[]{new PinningTrustManager()}, null);
        return sslContext.getSocketFactory();
    }

    /**
     * Custom trust manager for additional certificate validation
     */
    private class PinningTrustManager implements X509TrustManager {
        @Override
        public void checkClientTrusted(X509Certificate[] chain, String authType) 
            throws CertificateException {
            // Client certificates not used
        }

        @Override
        public void checkServerTrusted(X509Certificate[] chain, String authType) 
            throws CertificateException {
            // Perform standard certificate validation
            if (chain == null || chain.length == 0) {
                throw new CertificateException("Certificate chain is empty");
            }
            
            // Check certificate validity
            for (X509Certificate cert : chain) {
                cert.checkValidity();
            }
            
            // Additional custom validation can be added here
            // For example, checking certificate transparency logs
        }

        @Override
        public X509Certificate[] getAcceptedIssuers() {
            return new X509Certificate[0];
        }
    }

    /**
     * Strict hostname verifier
     */
    private class StrictHostnameVerifier implements HostnameVerifier {
        @Override
        public boolean verify(String hostname, SSLSession session) {
            // Verify hostname matches certificate
            try {
                Certificate[] certs = session.getPeerCertificates();
                if (certs.length > 0 && certs[0] instanceof X509Certificate) {
                    X509Certificate cert = (X509Certificate) certs[0];
                    
                    // Check if hostname matches certificate CN or SAN
                    String cn = getCN(cert);
                    if (cn != null && cn.equalsIgnoreCase(hostname)) {
                        return true;
                    }
                    
                    // Check Subject Alternative Names
                    return checkSAN(cert, hostname);
                }
            } catch (SSLPeerUnverifiedException e) {
                Log.e(TAG, "Hostname verification failed", e);
            }
            
            sendSecurityEvent("hostname_verification_failed", hostname);
            return false;
        }
        
        private String getCN(X509Certificate cert) {
            String dn = cert.getSubjectX500Principal().getName();
            for (String part : dn.split(",")) {
                if (part.trim().startsWith("CN=")) {
                    return part.trim().substring(3);
                }
            }
            return null;
        }
        
        private boolean checkSAN(X509Certificate cert, String hostname) {
            try {
                java.util.Collection<java.util.List<?>> sans = 
                    cert.getSubjectAlternativeNames();
                if (sans != null) {
                    for (java.util.List<?> san : sans) {
                        if (san.size() >= 2) {
                            Integer type = (Integer) san.get(0);
                            if (type == 2) { // DNS name
                                String dns = (String) san.get(1);
                                if (dns.equalsIgnoreCase(hostname)) {
                                    return true;
                                }
                            }
                        }
                    }
                }
            } catch (Exception e) {
                Log.e(TAG, "Error checking SAN", e);
            }
            return false;
        }
    }

    /**
     * Send security events to JavaScript
     */
    private void sendSecurityEvent(String eventType, String details) {
        WritableMap params = Arguments.createMap();
        params.putString("type", eventType);
        params.putString("details", details);
        params.putDouble("timestamp", System.currentTimeMillis());
        
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit("CertificatePinningSecurityEvent", params);
    }
}